File: algorithms/sort/comparison/cocktail_shaker/java/CocktailShaker.java
public class CocktailShaker {
    public void sort(int[] a) {
        var swaped = false;
        int start = 0, end = a.length;
        do {
            swaped = false;
            for(int i = start + 1; i < end; i++) {
                if(a[i-1] > a[i]) {
                    var tmp = a[i - 1];
                    a[i-1] = a[i];
                    a[i] = tmp;
                    swaped = true;
                }
            }
            if(swaped) swaped = false; 
            else break;
            for(int i = end - 1; i > start; i--) {
                if(a[i-1] > a[i]) {
                    var tmp = a[i - 1];
                    a[i-1] = a[i];
                    a[i] = tmp;
                    swaped = true;
                }
            }
            start++;
            end--;
        } while(swaped);
    }
}

File: algorithms/sort/comparison/cocktail_shaker/kotlin/CocktailShaker.kt
class CocktailShaker {
    fun sort(a: IntArray) {
       var swaped:Boolean
       var start = 0;
       var end = a.size - 1
       do {
           swaped = false
           for(i in (start + 1)..end) {
               if(a[i-1] > a[i]) {
                   var tmp = a[i-1]
                   a[i-1] = a[i]
                   a[i] = tmp
                   swaped = true
               }
           }
           if (swaped) swaped = false
           else break
           for(i in end downTo start + 1) {
               if(a[i-1] > a[i]) {
                   var tmp = a[i-1]
                   a[i-1] = a[i]
                   a[i] = tmp
                   swaped = true
               }
           }
           start++
           end--
       } while(swaped)
    }
}
